keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes's validation of IP addresses and CIDR strings (e.g.
+"`192.168.0.0/24`") in API fields has historically been too lax,
+accepting any string accepted by the underlying functions
+(`net.ParseIP` and `net.ParseCIDR`) even though in some cases these
+strings had ambiguous meanings that could lead to security problems.
+After [golang 1.17 changed the handling of IP addresses with leading
+"0"s] to avoid [CVE-2021-29923], there was consensus that we should
+eventually update API validation to be stricter as well, but we did
+not do this right away, because we didn't want to retroactively
+invalidate existing API objects. This KEP sets forth the plan for
+finally moving forward on that.
+
+[golang 1.17 changed the handling of IP addresses with leading "0"s]: https://go-review.googlesource.com/c/go/+/325829
+[CVE-2021-29923]: https://nvd.nist.gov/vuln/detail/cve-2021-29923
+
+## Motivation
+
+### Goals
+
+- Update the validation of all IP-valued and CIDR-valued API fields in
+  core (`k8s.io/api`) types to require stricter validation, to avoid
+  potential security problems with ambiguously-parsable IPs/CIDRs.
+
+    - Don't require pre-existing invalid fields to be fixed on
+      unrelated updates. E.g., if an existing Service has `clusterIP:
+      172.030.099.099`, allow a user to change the Service's
+      `selector` without being forced to fix the bad `clusterIP` at
+      the same time.
+
+    - Allow immutable fields to be fixed if they are invalid. E.g., if
+      an existing Service has `clusterIP: 172.030.099.099`, allow
+      changing it to `clusterIP: 172.30.99.99` even though `clusterIP`
+      is normally immutable.
+
+- Update the [CEL IP/CIDR validation helpers] to use the same code as
+  the new core API validation. (The CEL helpers already have the
+  correct semantics; this is just about not having two separate
+  implementations.)
+
+```
+<<[UNRESOLVED transition ]>>
+
+- MAYBE do a few releases where IP/CIDR values that are invalid
+  according to the new rules cause apiserver warnings, before actually
+  changing the validation.
```

Comment:
This is already done for Services https://github.com/kubernetes/kubernetes/pull/114505

Reply:
(yeah, that's what inspired this maybe-goal and the next one)